---
title: "HiCDOC analysis"
output: 
    rmarkdown::html_vignette:
       fig_width: 7
       fig_height: 5
vignette: >
  %\VignetteIndexEntry{HiCDOC analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{HiCDOC}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
#library(HiCDOC)
devtools::load_all(".")
```

HiCDOC normalizes intrachromosomal Hi-C matrices and detects A/B compartments
with multiple replicates using unsupervised learning.

HiCDOC is segmented in a collection of functions, which can be used to construct a
custom pipeline :

1. Normalize technical biases with using a cyclic loess<sup>[[publication][multihiccompare-publication]][[implementation][cyclic-loess-implementation]]</sup>
2. Normalize biological biases with Knight-Ruiz method<sup>[[publication][knight-ruiz-publication]][[implementation][knight-ruiz-implementation]]</sup>
3. Normalize distance effect with an MD loess
4. Detect compartments with constrained k-means<sup>[[publication][constrained-k-means-publication]][[implementation][constrained-k-means-implementation]]</sup>
5. Plot compartment changes and measures

## Start with HiCDOC

A small dataset is shipped with the package:

```{r}
object <- HiCDOCExample()
```

### Cleaning

First, the data should be cleaned, by filtering the small chromosomes  and removing the weak bins. 

Filtering the small chromosomes means removing the chromosomes that have less than a given number of bins.
The number of bins by default is 100. 

Removing the weak positions means identifying and removing the bins for which the mean value 
is lower of equal to a threshold, for at least one condition and one replicate. 

The values are then removed for all the chromosome (for all conditions and replicates) and kept in 
object@weakBins. 

```{r}
object  <- filterSmallChromosomes(object)
object  <- filterWeakPositions(object)
```

### Normalization

The data can be plotted (here for the 1st chromosome):

```{r}
plotInteractionMatrix(object, chromosomeId = 1)
```

(the rest of the vignette will focus on the 1st chromosome, for clarity reasons)

#### Technical bias
the normalization for technical bias can be performed:

```{r}
object <- normalizeTechnicalBiases(object)
plotInteractionMatrix(object, chromosomeId = 1)
```

#### Biological bias
the normalization for biological bias can be performed:

```{r}
object <- normalizeBiologicalBiases(object)
plotInteractionMatrix(object, chromosomeId = 1)
```

#### Distance effect

The normalization for distance effect can be performed and we look again at the distance effect:

```{r}
objectnormed <- normalizeDistanceEffect(object)
plotInteractionMatrix(objectnormed, chromosomeId = 1)
```

We can look at the distance effect before and after distance normalization: 

```{r}
plotDistanceEffect(object)
```
```{r}
plotDistanceEffect(objectnormed)
```

### Detect Compartments

On the normalized object, we can now run the detection of compartments, and plot them (here for the 1st chromosome):

```{r}
objectnormed <- detectCompartments(objectnormed)
```

```{r}
plotCompartmentChanges(objectnormed, chromosomeId = 1, padjThreshold = 0.05)
```

The objects created by the `detectCompartments` function (`differences`, `concordances`, `compartments`) can be viewed separately:

```{r} 
differences(objectnormed, pvalue = 0.1)
```
```{r}
concordances(objectnormed)
```
```{r}
compartments(objectnormed)
```

### Visualization

Various visualizations are also available:

```{r}
plotAB(objectnormed, chromosomeId = "17")
plotDiffConcordances(objectnormed)
plotCentroids(objectnormed, chromosomeId = 1)
```

## Input format

### Table format

Each script accepts a tab-separated multi-replicate sparse matrix with a header
line and optional comment lines.

    # Optional comment lines
    # ...
    chromosome    position 1    position 2    replicate 1.1    replicate 1.2    replicate 2.1    ...
    3             1500000       7500000       145              184              72               ...
    ...

The interaction proportions between `position 1` and `position 2` are reported
in each replicate column, named `replicate <condition.replicate>`. There is no
limit to the number of replicates and conditions.

### Cool files

Files with the [cool format](https://github.com/mirnylab/cooler/) can also be used.
Call the function `HiCDOCDataSetFromCool` this way:

```R
HiCDOCDataSetFromCool(coolFiles, replicates, conditions)
```

where:

 - `coolFiles` is a vector containing the files names
 - `replicates` is a vector containing the names of the replicates (such as `c("rep1_wt", "rep1_wt", ...)`)
 - `conditions` is a vector containing the id of each condition (such as `c(1, 1, 1, 2, 2, 2)` if you have 3 replicates for each condition).

Once your data is loaded, you can follow the usual pipe-line.

### Load from sparse matrix

Start you script with:
```R
object <- HiCDOCDataSetFromSparseMatrix(matrix)
```
Then, follow the usual pipe-line.

### Load from `.hic` files

If the `.hic` files are stored in the vector `hicFiles`, start your script with:

```R
resolution <- 100000     # set as desired
dataSet <- HiCDOCDataSetFromHic(hicFiles, replicates, conditions, resolution)
object  <- HiCDOCExp(dataSet)
```

## References

Philip A. Knight, Daniel Ruiz, A fast algorithm for matrix balancing, _IMA
Journal of Numerical Analysis_, Volume 33, Issue 3, July 2013, Pages 1029–1047,
https://doi.org/10.1093/imanum/drs019

Rajendra Kumar, Haitham Sobhy, Per Stenberg, Ludvig Lizana, Genome contact map
explorer: a platform for the comparison, interactive visualization and analysis
of genome contact maps, _Nucleic Acids Research_, Volume 45, Issue 17, 29
September 2017, Page e152, https://doi.org/10.1093/nar/gkx644

John C Stansfield, Kellen G Cresswell, Mikhail G Dozmorov, multiHiCcompare:
joint normalization and comparative analysis of complex Hi-C experiments,
_Bioinformatics_, 2019, https://doi.org/10.1093/bioinformatics/btz048

Kiri Wagstaff, Claire Cardie, Seth Rogers, Stefan Schrödl, Constrained K-means
Clustering with Background Knowledge, _Proceedings of 18th International
Conference on Machine Learning_, 2001, Pages 577-584,
https://pdfs.semanticscholar.org/0bac/ca0993a3f51649a6bb8dbb093fc8d8481ad4.pdf

[multihiccompare-publication]: https://doi.org/10.1093/bioinformatics/btz048
[multihiccompare-installation]: https://bioconductor.org/packages/release/bioc/html/multiHiCcompare.html
[gcmapexplorer-publication]: https://doi.org/10.1093/nar/gkx644
[gcmapexplorer-installation]: https://gcmapexplorer.readthedocs.io/en/latest/install.html
[orca-installation]: https://github.com/plotly/orca#installation
[cyclic-loess-implementation]: https://bioconductor.org/packages/release/bioc/vignettes/multiHiCcompare/inst/doc/multiHiCcompare.html#cyclic-loess-normalization
[knight-ruiz-publication]: https://doi.org/10.1093/imanum/drs019
[knight-ruiz-implementation]: https://gcmapexplorer.readthedocs.io/en/latest/commands/normKR.html
[rnr-implementation]: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.RadiusNeighborsRegressor.html
[interaction-mean-implementation]: https://gcmapexplorer.readthedocs.io/en/latest/commands/normMCFS.html
[constrained-k-means-publication]: https://pdfs.semanticscholar.org/0bac/ca0993a3f51649a6bb8dbb093fc8d8481ad4.pdf
[constrained-k-means-implementation]: https://github.com/Behrouz-Babaki/COP-Kmeans
[silhouette-implementation]: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_samples.html
