<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A/B compartment detection and differential analysis • HiCDOC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A/B compartment detection and differential analysis">
<meta property="og:description" content="
    HiCDOC normalizes intrachromosomal Hi-C matrices, uses unsupervised learning
    to predict A/B compartments from multiple replicates, and detects
    significant compartment changes between experiment conditions.
    It provides a collection of functions assembled into a pipeline to filter 
    and normalize the data, predict the compartments and visualize the results. 
    It accepts several type of data: tabular `.tsv` files, Cooler `.cool` or 
    `.mcool` files, Juicer `.hic` files or HiC-Pro `.matrix` and `.bed` files.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">HiCDOC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/HiCDOC.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mzytnicki/HiCDOC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="hicdoc-compartments-prediction-and-differential-analysis-with-multiple-replicates" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#hicdoc-compartments-prediction-and-differential-analysis-with-multiple-replicates" class="anchor"></a>HiCDOC: Compartments prediction and differential analysis with multiple replicates</h1></div>
<p>HiCDOC normalizes intrachromosomal Hi-C matrices, uses unsupervised learning to predict A/B compartments from multiple replicates, and detects significant compartment changes between experiment conditions.</p>
<p>It provides a collection of functions assembled into a pipeline:</p>
<ol>
<li>
<a href="#filtering-data">Filter</a>:
<ol>
<li>Remove chromosomes which are too small to be useful.</li>
<li>Filter sparse replicates to remove uninformative replicates with few interactions.</li>
<li>Filter positions (<em>bins</em>) which have too few interactions.</li>
</ol>
</li>
<li>
<a href="#normalizing-biases">Normalize</a>:
<ol>
<li>Normalize technical biases using <a href="https://doi.org/10.1093/bioinformatics/btz048">cyclic loess normalization</a>, so that matrices are comparable.</li>
<li>Normalize biological biases using <a href="https://doi.org/10.1093/imanum/drs019">Knight-Ruiz matrix balancing</a>, so that all the bins are comparable.</li>
<li>Normalize the distance effect, which results from higher interaction proportions between closer regions, with a MD loess.</li>
</ol>
</li>
<li>
<a href="#predicting-compartments-and-differences">Predict</a>:
<ol>
<li>Predict compartments using <a href="https://pdfs.semanticscholar.org/0bac/ca0993a3f51649a6bb8dbb093fc8d8481ad4.pdf">constrained K-means</a>.</li>
<li>Detect significant differences between experiment conditions.</li>
</ol>
</li>
<li>
<a href="#visualizing-data-and-results">Visualize</a>:
<ol>
<li>Plot the interaction matrices of each replicate.</li>
<li>Plot the overall distance effect on the proportion of interactions.</li>
<li>Plot the compartments in each chromosome, along with their concordance (confidence measure) in each replicate, and significant changes between experiment conditions.</li>
<li>Plot the overall distribution of concordance differences.</li>
<li>Plot the result of the PCA on the compartments’ centroids.</li>
<li>Plot the boxplots of self interaction ratios (differences between self interactions and the medians of other interactions) of each compartment, which is used for the A/B classification.</li>
</ol>
</li>
</ol>
</div>
<div id="table-of-contents" class="section level1">
<h1 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of contents</h1>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#quick-start">Quick Start</a></li>
<li>
<a href="#quick-start">Usage</a>
<ul>
<li>
<a href="#importing-hi-c-data">Importing Hi-C data</a>
<ul>
<li><a href="#tabular-files">Tabular files</a></li>
<li><a href="#cooler-files">Cooler files</a></li>
<li><a href="#juicer-files">Juicer files</a></li>
<li><a href="#hic-pro-files">HiC-Pro files</a></li>
</ul>
</li>
<li>
<a href="#running-the-hicdoc-pipeline">Running the HiCDOC pipeline</a>
<ul>
<li><a href="#filtering-data">Filtering data</a></li>
<li><a href="#normalizing-biases">Normalizing biases</a></li>
<li><a href="#predicting-compartments-and-differences">Predicting compartments and differences</a></li>
</ul>
</li>
<li><a href="#visualizing-data-and-results">Visualizing data and results</a></li>
</ul>
</li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install, execute the following commands in your console:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'install.packages("devtools")'</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'devtools::install_github("mzytnicki/HiCDOC")'</span></span></code></pre></div>
<p>After installation, the package can be loaded in R &gt;= 4.0:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mzytnicki/HiCDOC">"HiCDOC"</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<p>To try out HiCDOC, load the simulated toy data set:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">exampleHiCDOCDataSet</span><span class="op">)</span>
<span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="va">exampleHiCDOCDataSet</span></code></pre></div>
<p>Then run the default pipeline on the created object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOC.html">HiCDOC</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>And plot some results:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotCompartmentChanges.html">plotCompartmentChanges</a></span><span class="op">(</span><span class="va">hic.experiment</span>, chromosome <span class="op">=</span> <span class="st">'Y'</span><span class="op">)</span></code></pre></div>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<div id="importing-hi-c-data" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-hi-c-data" class="anchor"></a>Importing Hi-C data</h2>
<p>HiCDOC can import Hi-C data sets in various different formats:</p>
<ul>
<li>Tabular <code>.tsv</code> files.</li>
<li>Cooler <code>.cool</code> or <code>.mcool</code> files.</li>
<li>Juicer <code>.hic</code> files.</li>
<li>HiC-Pro <code>.matrix</code> and <code>.bed</code> files.</li>
</ul>
<div id="tabular-files" class="section level3">
<h3 class="hasAnchor">
<a href="#tabular-files" class="anchor"></a>Tabular files</h3>
<p>A tabular file is a tab-separated multi-replicate sparse matrix with a header:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>chromosome    position <span class="dv">1</span>    position <span class="dv">2</span>    C1.R1    C1.R2    C2.R1    ...</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>             <span class="dv">1500000</span>       <span class="dv">7500000</span>       <span class="dv">145</span>      <span class="dv">184</span>      <span class="dv">72</span>       ...</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>...</span></code></pre></div>
<p>The interaction proportions between <code>position 1</code> and <code>position 2</code> of <code>chromosome</code> are reported in each <code>condition.replicate</code> column. There is no limit to the number of conditions and replicates.</p>
<p>To load Hi-C data in this format:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOCDataSetFromTabular.html">HiCDOCDataSetFromTabular</a></span><span class="op">(</span><span class="st">'path/to/data.tsv'</span><span class="op">)</span></code></pre></div>
</div>
<div id="cooler-files" class="section level3">
<h3 class="hasAnchor">
<a href="#cooler-files" class="anchor"></a>Cooler files</h3>
<p>To load <code>.cool</code> or <code>.mcool</code> files generated by <a href="https://cooler.readthedocs.io/en/latest/">Cooler</a>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Path to each file</span>
<span class="va">paths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">'path/to/condition-1.replicate-1.cool'</span>,
  <span class="st">'path/to/condition-1.replicate-2.cool'</span>,
  <span class="st">'path/to/condition-2.replicate-1.cool'</span>,
  <span class="st">'path/to/condition-2.replicate-2.cool'</span>,
  <span class="st">'path/to/condition-3.replicate-1.cool'</span>
<span class="op">)</span>

<span class="co"># Replicate and condition of each file. Can be names instead of numbers.</span>
<span class="va">replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>

<span class="co"># Resolution to select in .mcool files</span>
<span class="va">binSize</span> <span class="op">=</span> <span class="fl">500000</span>

<span class="co"># Instantiation of data set</span>
<span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOCDataSetFromCool.html">HiCDOCDataSetFromCool</a></span><span class="op">(</span>
  <span class="va">paths</span>,
  replicates <span class="op">=</span> <span class="va">replicates</span>,
  conditions <span class="op">=</span> <span class="va">conditions</span>,
  binSize <span class="op">=</span> <span class="va">binSize</span> <span class="co"># Specified for .mcool files.</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="juicer-files" class="section level3">
<h3 class="hasAnchor">
<a href="#juicer-files" class="anchor"></a>Juicer files</h3>
<p>To load <code>.hic</code> files generated by <a href="https://github.com/aidenlab/juicer/wiki/Data">Juicer</a>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Path to each file</span>
<span class="va">paths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">'path/to/condition-1.replicate-1.hic'</span>,
  <span class="st">'path/to/condition-1.replicate-2.hic'</span>,
  <span class="st">'path/to/condition-2.replicate-1.hic'</span>,
  <span class="st">'path/to/condition-2.replicate-2.hic'</span>,
  <span class="st">'path/to/condition-3.replicate-1.hic'</span>
<span class="op">)</span>

<span class="co"># Replicate and condition of each file. Can be names instead of numbers.</span>
<span class="va">replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>

<span class="co"># Resolution to select</span>
<span class="va">binSize</span> <span class="op">&lt;-</span> <span class="fl">500000</span>

<span class="co"># Instantiation of data set</span>
<span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOCDataSetFromHiC.html">HiCDOCDataSetFromHiC</a></span><span class="op">(</span>
  <span class="va">paths</span>,
  replicates <span class="op">=</span> <span class="va">replicates</span>,
  conditions <span class="op">=</span> <span class="va">conditions</span>,
  binSize <span class="op">=</span> <span class="va">binSize</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="hic-pro-files" class="section level3">
<h3 class="hasAnchor">
<a href="#hic-pro-files" class="anchor"></a>HiC-Pro files</h3>
<p>To load <code>.matrix</code> and <code>.bed</code> files generated by <a href="https://github.com/nservant/HiC-Pro">HiC-Pro</a>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Path to each matrix file</span>
<span class="va">matrixPaths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">'path/to/condition-1.replicate-1.matrix'</span>,
  <span class="st">'path/to/condition-1.replicate-2.matrix'</span>,
  <span class="st">'path/to/condition-2.replicate-1.matrix'</span>,
  <span class="st">'path/to/condition-2.replicate-2.matrix'</span>,
  <span class="st">'path/to/condition-3.replicate-1.matrix'</span>
<span class="op">)</span>

<span class="co"># Path to each bed file</span>
<span class="va">bedPaths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">'path/to/condition-1.replicate-1.bed'</span>,
  <span class="st">'path/to/condition-1.replicate-2.bed'</span>,
  <span class="st">'path/to/condition-2.replicate-1.bed'</span>,
  <span class="st">'path/to/condition-2.replicate-2.bed'</span>,
  <span class="st">'path/to/condition-3.replicate-1.bed'</span>
<span class="op">)</span>

<span class="co"># Replicate and condition of each file. Can be names instead of numbers.</span>
<span class="va">replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>

<span class="co"># Instantiation of data set</span>
<span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOCDataSetFromHiCPro.html">HiCDOCDataSetFromHiCPro</a></span><span class="op">(</span>
  matrixPaths <span class="op">=</span> <span class="va">matrixPaths</span>,
  bedPaths <span class="op">=</span> <span class="va">bedPaths</span>,
  replicates <span class="op">=</span> <span class="va">replicates</span>,
  conditions <span class="op">=</span> <span class="va">conditions</span>
<span class="op">)</span></code></pre></div>
</div>
</div>
<div id="running-the-hicdoc-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-hicdoc-pipeline" class="anchor"></a>Running the HiCDOC pipeline</h2>
<p>Once your data is loaded, you can run all the filtering, normalization, and prediction steps with:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/HiCDOC.html">HiCDOC</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>This one-liner runs all the steps detailed below.</p>
<div id="filtering-data" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-data" class="anchor"></a>Filtering data</h3>
<p>Remove small chromosomes of length smaller than 100 positions:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterSmallChromosomes.html">filterSmallChromosomes</a></span><span class="op">(</span><span class="va">hic.experiment</span>, threshold <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>Remove sparse replicates filled with less than 30% non-zero interactions:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterSparseReplicates.html">filterSparseReplicates</a></span><span class="op">(</span><span class="va">hic.experiment</span>, threshold <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></code></pre></div>
<p>Remove weak positions with less than 1 interaction in average:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/filterWeakPositions.html">filterWeakPositions</a></span><span class="op">(</span><span class="va">hic.experiment</span>, threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
</div>
<div id="normalizing-biases" class="section level3">
<h3 class="hasAnchor">
<a href="#normalizing-biases" class="anchor"></a>Normalizing biases</h3>
<p>Normalize technical biases such as sequencing depth:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/normalizeTechnicalBiases.html">normalizeTechnicalBiases</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>Normalize biological biases (such as GC content, number of restriction sites, etc.):</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/normalizeBiologicalBiases.html">normalizeBiologicalBiases</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>Normalize the distance effect resulting from higher interaction proportions between closer regions:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/normalizeDistanceEffect.html">normalizeDistanceEffect</a></span><span class="op">(</span><span class="va">hic.experiment</span>, loessSampleSize <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></code></pre></div>
</div>
<div id="predicting-compartments-and-differences" class="section level3">
<h3 class="hasAnchor">
<a href="#predicting-compartments-and-differences" class="anchor"></a>Predicting compartments and differences</h3>
<p>Predict A and B compartments and detect significant differences:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hic.experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/detectCompartments.html">detectCompartments</a></span><span class="op">(</span>
  <span class="va">hic.experiment</span>,
  kMeansDelta <span class="op">=</span> <span class="fl">0.0001</span>,
  kMeansIterations <span class="op">=</span> <span class="fl">50</span>,
  kMeansRestarts <span class="op">=</span> <span class="fl">20</span>
<span class="op">)</span></code></pre></div>
</div>
</div>
<div id="visualizing-data-and-results" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-data-and-results" class="anchor"></a>Visualizing data and results</h2>
<p>Plot the interaction matrix of each replicate:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotInteractions.html">plotInteractions</a></span><span class="op">(</span><span class="va">hic.experiment</span>, chromosome <span class="op">=</span> <span class="st">'3'</span><span class="op">)</span></code></pre></div>
<p>Plot the overall distance effect on the proportion of interactions:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotDistanceEffect.html">plotDistanceEffect</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>List and plot compartments with their concordance (confidence measure) in each replicate, and significant changes between experiment conditions:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/HiCDOCDataSet-methods.html">compartments</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span>

<span class="fu"><a href="reference/HiCDOCDataSet-methods.html">concordances</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span>

<span class="fu"><a href="reference/HiCDOCDataSet-methods.html">differences</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span>

<span class="fu"><a href="reference/plotCompartmentChanges.html">plotCompartmentChanges</a></span><span class="op">(</span><span class="va">hic.experiment</span>, chromosome <span class="op">=</span> <span class="st">'3'</span><span class="op">)</span></code></pre></div>
<p>Plot the overall distribution of concordance differences:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotConcordanceDifferences.html">plotConcordanceDifferences</a></span><span class="op">(</span><span class="va">hic.experiment</span><span class="op">)</span></code></pre></div>
<p>Plot the result of the PCA on the compartments’ centroids:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotCentroids.html">plotCentroids</a></span><span class="op">(</span><span class="va">hic.experiment</span>, chromosome <span class="op">=</span> <span class="st">'3'</span><span class="op">)</span></code></pre></div>
<p>Plot the boxplots of self interaction ratios (differences between self interactions and the median of other interactions) of each compartment:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/plotSelfInteractionRatios.html">plotSelfInteractionRatios</a></span><span class="op">(</span><span class="va">hic.experiment</span>, chromosome <span class="op">=</span> <span class="st">'3'</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>John C Stansfield, Kellen G Cresswell, Mikhail G Dozmorov, multiHiCcompare: joint normalization and comparative analysis of complex Hi-C experiments, <em>Bioinformatics</em>, 2019, <a href="https://doi.org/10.1093/bioinformatics/btz048" class="uri">https://doi.org/10.1093/bioinformatics/btz048</a></p>
<p>Philip A. Knight, Daniel Ruiz, A fast algorithm for matrix balancing, <em>IMA Journal of Numerical Analysis</em>, Volume 33, Issue 3, July 2013, Pages 1029–1047, <a href="https://doi.org/10.1093/imanum/drs019" class="uri">https://doi.org/10.1093/imanum/drs019</a></p>
<p>Kiri Wagstaff, Claire Cardie, Seth Rogers, Stefan Schrödl, Constrained K-means Clustering with Background Knowledge, <em>Proceedings of 18th International Conference on Machine Learning</em>, 2001, Pages 577-584, <a href="https://pdfs.semanticscholar.org/0bac/ca0993a3f51649a6bb8dbb093fc8d8481ad4.pdf" class="uri">https://pdfs.semanticscholar.org/0bac/ca0993a3f51649a6bb8dbb093fc8d8481ad4.pdf</a></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mzytnicki/HiCDOC/">https://​github.com/​mzytnicki/​HiCDOC/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mzytnicki/HiCDOC/issues">https://​github.com/​mzytnicki/​HiCDOC/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>LGPL (&gt;= 3)</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Kurylo Cyril <br><small class="roles"> Author </small>  </li>
<li>Zytnicki Matthias <br><small class="roles"> Author </small>  </li>
<li>Foissac Sylvain <br><small class="roles"> Author </small>  </li>
<li>Maigné Élise <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Kurylo Cyril, Zytnicki Matthias, Foissac Sylvain, Maigné Élise.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
